tools/serverpod_cli/lib/src/util/pubspec_lock_parser.dart

```diff
@@ -0,0 +1,252 @@
+import 'dart:io';
+
+import 'package:pub_semver/pub_semver.dart';
+import 'package:pubspec_parse/pubspec_parse.dart';
+import 'package:yaml/yaml.dart';
+
+/// Represents the type of a dependency in pubspec.lock
+enum DependencyType {
+  /// Direct dependency from pubspec.yaml
+  directMain('direct main'),
+
+  /// Direct dev dependency from pubspec.yaml
+  directDev('direct dev'),
+
+  /// Transitive dependency (dependency of a dependency)
+  transitive('transitive'),
+
+  /// Direct overridden dependency
+  directOverridden('direct overridden');
+
+  final String value;
+  const DependencyType(this.value);
+
+  static DependencyType fromString(String value) {
+    return DependencyType.values.firstWhere(
+      (type) => type.value == value,
+      orElse: () => throw ArgumentError('Unknown dependency type: $value'),
+    );
+  }
+}
+
+/// Represents a package entry in pubspec.lock
+class LockedPackage {
+  final String name;
+  final Dependency dependency;
+  final DependencyType dependencyType;
+  final Version version;
+  final String? sha256;
+  final String source;
+
+  LockedPackage({
+    required this.name,
+    required this.dependency,
+    required this.dependencyType,
+    required this.version,
+    this.sha256,
+    required this.source,
+  });
+
+  /// Whether this is a direct dependency (not transitive)
+  bool get isDirect =>
+      dependencyType == DependencyType.directMain ||
+      dependencyType == DependencyType.directDev ||
+      dependencyType == DependencyType.directOverridden;
+
+  /// Whether this is a dev dependency
+  bool get isDevDependency => dependencyType == DependencyType.directDev;
+
+  /// Whether this is a transitive dependency
+  bool get isTransitive => dependencyType == DependencyType.transitive;
+}
+
+/// Parser for pubspec.lock files
+class PubspecLockParser {
```

Comment:
Did you consider using an existing package such as [pubspec_lock(https://pub.dev/packages/pubspec_lock)? If so, what made you roll your own?

Reply:
I did look at it. I felt the API weird, and also it gives packages as a list which I felt was better as a map instead. Since it's not very complicated to parse, I opted for implementing a simpler API backed by a map instead. Do you feel we should use this package instead?